### PR TITLE
Ddo 1519 create migrations environment manager

### PR DESCRIPTION
**Pull request by jyang-broad**
Lots of new things here:
Added a bunch of migrations! Not all of them are in use yet, but wanted to get code PR'd before it gets too :chonk:

Main Change:  Added Environments

Notes:
- this is a POC PR for Environments there are  fields that haven't been dealt with but want to sanity check that I'm going in the right direction.
- Went the route of not mocking EnvironmentController for tests, opinions here are welcome, not sure how I want to initialize the testApp env.
- Moved CreateEnvironmentRequest into model.db, thought is that currently the persistance layer for our objects are strictly DB bound, so the model-layer should own the struct that it accepts as "good enough" for the db.
- my pointer game sucks, still figuring that stuff out.
- I can work on handler/api for testing if that's useful, but I want to test logic on the controller level since I expect that there will be other parts of code that want to call list/get/create/etc, and API tests I expect to be more about marshalling/unmarshalling.
- AFAIK postgres does not add FKey indexes by default for various corner-casey reasons, but I think it's generally accepted to be good in most cases.

### CODE REVIEW

File: internal/sherlock/sherlock_integration_test.go [hunk 1]
```diff
@@ -446,7 +446,9 @@ func integrationSetup(t *testing.T) {
 	// when running tests workdir is the package directory ie cmd/server
 	// so a relative path to changelogs is needed.
 	// TODO cleaner method to supply path to changelogs and run migration in tests
-	if err := db.ApplyMigrations("../../db/migrations", sherlock.Config); err != nil {
+	if err := db.ApplyMigrations("../../db/migrations", sherlock.Config); err == migrate.ErrNoChange {
```
Comment (mflinn-broad) [inline internal/sherlock/sherlock_integration_test.go:449]: This comment from the linter is what is causing all your compilation failures. Seems like a missing import statement.. I'd like to see the CI checks passing before doing a full review.

**Comment by mflinn-broad** [comment]
> Went the route of not mocking EnvironmentController for tests, opinions here are welcome, not sure how I want to initialize the testApp env.

To clarify the testing strategy in some of the other packages. I'm never mocking the controllers. In the unit tests I do mock the persistence layer ie the `store`. If you check out some of the other package test code it is actually testing controller behavior in addition to just marshaling/ unmarshaling , ie there a number of assertions that verify certain methods are being invoked by the controller or not based on circumstances. You can verify this by looking at the coverage report from the last merged pr. 

I probably should rename the files to something besides handlers_test.go since they are actually testing both handler and controller behavior

File: internal/environments/environments_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	Config = viper.New()
+)
+
+type TestApplication struct {
+	Environments *EnvironmentController
+	DB           *gorm.DB
+}
+
+// New returns a new instance of the core sherlock application
+func initTestApp() *TestApplication {
```
Comment (mflinn-broad) [inline internal/environments/environments_test.go:58]: If you are going to write tests that expect a real postgres instance can you include [this](https://github.com/broadinstitute/sherlock/blob/72cc5d0157c3e9a246786f403ee13cc11ea7aeb6/internal/sherlock/sherlock_integration_test.go#L438) in the setup code. There are points in the ci pipeline and the make tooling that only run the unit tests (those without external dependencies) for example the image build ci job. This will enable these tests to be skipped when there isn't an actual postgres instance available.

File: internal/environments/environments_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+	//"github.com/broadinstitute/sherlock/internal/sherlock"
+	"github.com/broadinstitute/sherlock/internal/db"
+	"github.com/broadinstitute/sherlock/internal/tools"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+	"gorm.io/gorm"
+)
+
+func TestCreateEnvironments(t *testing.T) {
```
Comment (mflinn-broad) [inline internal/environments/environments_test.go:17]: since this is by our definition an integration Test (it uses a real postgres instance). Can you add `Integration` somewhere in the name of the test function. The make tooling looks for this to selectively run just integration tests.

File: internal/environments/environments.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import (
+	"errors"
+
+	"gorm.io/gorm"
+)
+
+// EnvironmentController is the management layer for environments
+type EnvironmentController struct {
+	store environmentStore
+	// store jobRequestStore // TODO: placeholder for other store environment controller can touch
```
Comment (mflinn-broad) [inline internal/environments/environments.go:18]: Small note if we ever need to give the environements controller access to other functionality, ie services. It might be better to embed the `ServiceController` rather than the `serviceStore`. That means we can only uses Services public API from other packages. This forces all the underlying logic for doing things related to services to stay in package services.

Comment (mflinn-broad) [reply internal/environments/environments.go:18]: ^ Is what I did for builds to give it the ability to create a new service if it recieves a build request for a service that doesn't exist already.

File: internal/sherlock/sherlock_integration_test.go (hunk 1 above)
Comment (jyang-broad) [reply internal/sherlock/sherlock_integration_test.go:449]: yep, thanks! Sorry I got my test to pass in vscode, forgotting to run the full suite and then just went punted debugging till today.

**Comment by mflinn-broad** [comment]
The actual code you have looks solid to me! My comments are mostly about test strategy and higher level organization

File: db/migrations/000010_add_fields_to_environments.up.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
+ALTER TABLE environments
+ADD COLUMN destroyed_at timestamp WITH TIME ZONE,
+ADD COLUMN is_permanent boolean,
+ADD COLUMN requester text,
+ADD COLUMN namespace text,
```
Comment (choover-broad) [inline db/migrations/000010_add_fields_to_environments.up.sql:5]: Hmmm, are we going to abandon the existing namespace convention? (Every Terra environment is deployed to a namespace `terra-ENVNAME`)

**Review by choover-broad — APPROVED** [review]
Exciting progress!

I don't have much to say about the code organization, will leave that to Mike.

I am a bit curious about some of the features / schema design, that's where most of my comments come from. They are left in the spirit of "perhaps this is a useful question to ask, perhaps not." Not a blocker for merging or continuing work towards the preview environments MVP.

File: db/migrations/000009_create_roles_table.down.sql [hunk 6]
```diff
@@ -0,0 +1,2 @@
+
+DROP TABLE IF EXISTS roles;
```
[the remark this answers is not in the document]
Comment (mflinn-broad) [reply db/migrations/000009_create_roles_table.down.sql:2]: +1 although if we want to leave the table hanging around I'm fine with it 🤷🏻

[the remark this answers is not in the document]
Comment (choover-broad) [reply db/migrations/000009_create_roles_table.down.sql:2]: Hope we can apply YAGNI here and avoid having partially-implemented features lying around

File: db/migrations/000007_create_clusters_table.up.sql [hunk 7]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS clusters(
+    id                        serial PRIMARY KEY,
+    name                      text UNIQUE NOT NULL,
+    google_project            text UNIQUE,
+    service_account_token_url text,
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000007_create_clusters_table.up.sql:5]: I was imagining that we have different SA/Keys in vault for each cluster for permissions reasons. i.e. for live environments, we might not want Sherlock having full access, but we do want that ability for qa/preview/any new clusters.

[the remark this answers is not in the document]
Comment (mflinn-broad) [reply db/migrations/000007_create_clusters_table.up.sql:5]: For live environments we don't want sherlock to have any access :)

File: db/migrations/000010_add_fields_to_environments.up.sql [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+ADD COLUMN customization jsonb,
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:6]: that's for the customizations per preview environment, like the unique name, specfic chart versions for apps.

File: db/migrations/000010_add_fields_to_environments.up.sql (hunk 5 above)
Comment (jyang-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:5]: I can remove that, thats probably an artifact from before we nailed down naming. I think we left off w/ terra-Environment.name

File: db/migrations/000007_create_clusters_table.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (choover-broad) [reply db/migrations/000007_create_clusters_table.up.sql:5]: We're deploying manifests via ArgoCD, right? 🤔 I don't think there's any reason for Sherlock to directly interact with the cluster API

File: db/migrations/000010_add_fields_to_environments.up.sql (hunk 5 above)
Comment (choover-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:5]: gotcha, tx!

File: db/migrations/000007_create_clusters_table.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (mflinn-broad) [reply db/migrations/000007_create_clusters_table.up.sql:5]: Yeah I don't understand why sherlock needs an SA to access the cluster either

File: db/migrations/000006_create_environment_reservations_table.up.sql [hunk 9]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS environment_reservations(
+    id                serial PRIMARY KEY,
+    environment_id    integer REFERENCES environments (id),
+    name              text,
+    source            text,
+    url               text,
+    ttl               interval, 
+    mutex             boolean,
+    metadata          jsonb,
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:9]: I can remove this, this was an artifact from when this was still job_requests and the table had larger scope.

File: db/migrations/000010_add_fields_to_environments.up.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mflinn-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:6]: Is an unstructured json blob the right way to handler that type of thing? 🤔

[the remark this answers is not in the document]
Comment (choover-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:6]: ah gotcha! Thanks.

In the spirit of offering EVEN MORE unsolicited opinions I wonder if there's merit in formally modeling this in the DB and API rather than allowing unstructured data. Just because:
* we're less likely to end up with bad data in the db from a misconfigured client
* there's a well-defined flow for making schema changes

File: db/migrations/000006_create_environment_reservations_table.up.sql [hunk 10]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS environment_reservations(
+    id                serial PRIMARY KEY,
+    environment_id    integer REFERENCES environments (id),
+    name              text,
+    source            text,
+    url               text,
+    ttl               interval,
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:7]: Totally agree on crutch behavior, there's also a world where we can use the TTL as informational and just bug people about a "stale" test if they haven't released the environment after the TTL. Either way I'm okay w/ the explicit release behavior.

File: db/migrations/000006_create_environment_reservations_table.up.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:9]: actaully I think there are two parts of data here. since there's intent to reuse environments ala allocation pools, there's two things pieces of data we have here, Environment.customization would show the current state of the environement, metadata would show the desired state of the environment ala all fiabs getting nuked and repopulated with expected app versions every run.

File: db/migrations/000006_create_environment_reservations_table.up.sql [hunk 11]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE IF NOT EXISTS environment_reservations(
+    id                serial PRIMARY KEY,
+    environment_id    integer REFERENCES environments (id),
+    name              text,
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:4]: name I expect to be some like `swatomation run #4452` source I expect to be the sourceIP/dns that the call is coming from. Ideally we'd see reservation locks/releases with pair name/source, but maybe if something breaks or gets stale we might see `devops cleanup` and `localhost` releases or something

[the remark this answers is not in the document]
Comment (choover-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:4]: Ah gotcha, thx for clarifying! 👍 

It's a bit tricky because while Jenkins has a concept of a user-friendly job run name, I don't know if GitHub actions does? The use of `name` here makes me double-take a bit, in lots of other places in Sherlock (services, environments) it's a strictly-formatted unique identifier whereas in this case it's more freeform and informational.

File: db/migrations/000010_add_fields_to_environments.up.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000010_add_fields_to_environments.up.sql:6]: I'm totally pro structured-api! I don't want people just throw blobs into our database, I'm fully expecting this column to be some sort of a rendered json created by us via api parameters or something we control.

I'm hesitant to formally model columns for a couple scenarios
a) I'm expecting the api to control/sanitize the input into the db, so I'm expecting that to deal with client-sourced problems.
b) if there's a new chart that gets added to helm, we would have to add a new row to give it a version, at the same time, deprecated services will end up with permanently dead rows. db migrations are scary and not something I'd like to do often or under external pressure (oh hey we have a new service, can you support it for us?).
c) I'm not clear atm what the scope of various customizations are for environments, essentially they're just helm charts overriding base configs afaik, so we might find this-or-that to be useful to tweak for various reasons (i dunno having a perf test have extra pods). Having to continuously think about db migrations/rollbacks every time we want to try something would just make it really hard.


Sorry if I'm misunderstanding something!

File: db/migrations/000007_create_clusters_table.up.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000007_create_clusters_table.up.sql:5]: uh right, argocd exists. blargh, this was probably me just blind copying from my schema, thanks for catching this!

File: db/migrations/000006_create_environment_reservations_table.up.sql (hunk 9 above)
[the remark this answers is not in the document]
Comment (choover-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:9]: >  Environment.customization would show the current state of the environement, metadata would show the desired state of the environment ala all fiabs getting nuked and repopulated with expected app versions every run

Having trouble wrapping my head around the distinction here but this is a really great point.

It sounds like we want to capture:
* **actual deployed versions**. For this it would probably make sense to re-use the Builds/Deploys endpoints Mike is implementing for lead time tracking, although we didn't initially plan on using those to track PR builds and deploys 
* **desired versions**. In other words, "preview env 1234 should deploy Leo off `my-feature-branch` in the Leo repo." The trick here is that _how_  those desired versions get wired in to Helm/ArgoCD seems to be really up in the air. Wondering if makes to omit this from this PR until there's more clarity. Especially since ArgoCD is developing this useful tool https://argocd-image-updater.readthedocs.io/en/stable/ which could change how we plug things in.

We *for sure* know that environment soft deletes and allocation pools are needed, and those seem like big/important enough features for one PR haha

File: db/migrations/000006_create_environment_reservations_table.up.sql [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+    source            text,
+    url               text,
+    ttl               interval, 
+    mutex             boolean,
+    metadata          jsonb,
+    created_at        timestamp WITH TIME ZONE default now(),
+    updated_at        timestamp WITH TIME ZONE default now()
```
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:11]: I'm treating environment_reservations as a time-series/events table, so it's not technically holding state, it's just recording when someone has locked/released a environment so.
`SELECT * FROM Environment_table WHERE environment == environment.id ORDER_BY created_at DESC LIMIT 1`
should always give you the state (either it's a release, or lock has a TTL that is either expired or not).

In the case that we don't respect TTL at all or just don't have it, then if the most recent row is a `lock` then it's locked and it would have been released at whenever we see a release event.

File: db/migrations/000006_create_environment_reservations_table.up.sql (hunk 11 above)
[the remark this answers is not in the document]
Comment (jyang-broad) [reply db/migrations/000006_create_environment_reservations_table.up.sql:4]: yeah the comment made me pause, I think it made a bunch more sense when it was job_request.name, but either way it's mean to link the reservation to some job. just looked into GHR, looks like there def are action_name + run_number ids, hopefully we can pull those.
Screen Shot 2021-09-17 at 12 45 23 PM

File: internal/environments/models.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+		return environment, err
+	}
+	return environment, nil
+}
+
+// get a Environment by name or err otherwise
+func (db dataStore) getByName(name string) (Environment, error) {
+	environment := Environment{}
+
+	if err := db.First(environment, name).Error; err != nil {
```
Comment (mflinn-broad) [inline internal/environments/models.go:106]:
```suggestion
	if err := db.First(environment, name).Error; err != nil {
```
```suggestion
	if err := db.Where(&Environment{Name: name}).First(&environment).Error; err != nil {
```

Sorry random comment I just stumbled across this. I think that syntax is some syntactic sugar that only works for get by ID. I think other look ups need to be done like this.

**Comment by jyang-broad** [comment]
WOOO a force push, I've removed environment_reservations for now as well as roles, and made the requested changes and well and fixed the breaking integration tests. If this is okay to merge, I will create another branch/PR to flesh out the current integration tests and move alot of the setup to `testing` package.

**Review by mflinn-broad — APPROVED** [review]
I checked out your branch and poked around a bit, everything looks reasonable to me with the exception of [this one bug](https://github.com/broadinstitute/sherlock/pull/11#discussion_r713403742) I mentioned yesterday. 

Approving though, not going to block over one minor issue :)
